nupic/research/frameworks/dendrites/routing/hardcoded.py

```diff
@@ -0,0 +1,140 @@
+                            module in addition to a linear feed-forward module
+    :param context_weights_fn: a function that returns a 3D torch Tensor that gives the
+                               near-optimal dendrite values for the specified
+                               dendrite_module, and has parameters `output_masks`,
+                               `context_vectors`, and `num_dendrites`
+    :param batch_size: the number of test inputs
+    """
+
+    # Initialize routing function that this task will try to hardcode
+    r = RoutingFunction(d_in=d_in, d_out=d_out, k=k, sparsity=0.7)
+
+    # Initialize context vectors, where each context vector corresponds to an output
+    # mask in the routing function
+    context_vectors = generate_context_vectors(
+        num_contexts=k,
+        n_dim=d_context,
+        percent_on=0.2
+    )
+
+    # Initialize dendrite module using the same feed-forward sparse weights as the
+    # routing function; also note that the value passed to `dendrite_sparsity` is
+    # irrelevant since the context weights are subsequently overwritten
+    dendritic_network = dendrite_module(
+        module=r.sparse_weights.module,
+        num_segments=k,
+        dim_context=d_context,
+        module_sparsity=0.7,
+        dendrite_sparsity=0.0
+    )
+
+    dendritic_network.register_buffer(
+        "zero_mask",
+        copy.deepcopy(r.sparse_weights.zero_mask.half())
+    )
+
+    # Choose the context weights specifically so that they can gate the outputs of the
+    # forward module
+    if context_weights_fn is not None:
+        dendritic_network.segments.weights.data = context_weights_fn(
+            output_masks=r.output_masks, context_vectors=context_vectors,
+            num_dendrites=k
+        )
+
+    # Sample a random batch of inputs and random batch of context vectors, and perform
+    # hardcoded routing test
+    x_test = 4.0 * torch.rand((batch_size, d_in)) - 2.0  # sampled i.i.d. from U[-2, 2)
+    context_inds_test = randint(low=0, high=k, size=batch_size).tolist()
+    context_test = torch.stack(
+        [context_vectors[j, :] for j in context_inds_test], dim=0
+    )
+
+    target = r(context_inds_test, x_test)
+    actual = dendritic_network(x_test, context_test)
+
+    # Report mean absolute error
+    mean_abs_error = torch.abs(target - actual).mean().item()
+    print(" Results over {} examples with {} random contexts".format(batch_size, k))
+    print(" > Mean absolute error between R(j, x) and output from dendritic network:")
+    print(" > \t{}".format(mean_abs_error))
+
+
+if __name__ == "__main__":
+
+    # Run the hardcoded routing test with a dendritic network that uses "dendrites as
+    # gating"; this achieves near-zero mean absolute error, which strongly suggests
+    # that dendritic networks can successfully route
+    d_in = 100
+    d_out = 100
+    num_contexts = 10
+    d_context = 100
+
+    run_hardcoded_routing_test(
```

Comment:
Create a unit test out of this, ensure the returned mean_abs_error is within some epsilon. This way we can test that changes to our dendrites will still pass this test.

Reply:
I added a single test case that tests whether mean_abs_error < epsilon for epsilon = 0.01.